kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -1128,8 +1056,13 @@ impl<'tcx> GotocCtx<'tcx> {
                 // Cast to a slice fat pointer.
                 assert_eq!(src_elt_type, dst_elt_type);
                 let dst_goto_len = self.codegen_const(*src_elt_count, None);
-                let dst_data_expr =
-                    src_goto_expr.cast_to(self.codegen_ty(*src_elt_type).to_pointer());
+                let src_pointee_ty = pointee_type(coerce_info.src_ty).unwrap();
+                let dst_data_expr = if src_pointee_ty.is_array() {
+                    src_goto_expr.cast_to(self.codegen_ty(*src_elt_type).to_pointer())
+                } else {
+                    // A struct that contains the type being coerced to a slice.
+                    src_goto_expr.cast_to(dst_data_type)
+                };
```

Comment:
A comment explaining this using C like pseudocode might make it easier to follow